### PR TITLE
Add media download operation.

### PR DESCRIPTION
**Pull request by dewniMW**
Issue: https://github.com/wso2/product-is/issues/8499

### CODE REVIEW

File: components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/common/MediaEndpointDataHolder.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    /**
+     * Get RealmService osgi service.
+     *
+     * @return RealmService
+     */
+    public RealmService getRealmService() {
+
+        return MediaEndpointDataHolder.getInstance().realmService;
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/common/MediaEndpointDataHolder.java:45]: no need to calll MediaEndpointDataHolder.getInstance() here.. we can directly return realmService

File: components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/common/MediaEndpointDataHolder.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    /**
+     * Set RealmService osgi service.
+     *
+     * @param realmService RealmService.
+     */
+    public void setRealmService(RealmService realmService) {
+
+        MediaEndpointDataHolder.getInstance().realmService = realmService;
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/common/MediaEndpointDataHolder.java:55]: change to this.realmService = realmService

File: components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/common/factory/RealmServiceFactory.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    public Class<RealmService> getObjectType() {
+
+        return RealmService.class;
+    }
+
+    @Override
+    protected RealmService createInstance() throws Exception {
+
+        if (this.realmService == null) {
+            RealmService realmService = (RealmService)
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/common/factory/RealmServiceFactory.java:41]: can't we inline line 41 to 43

File: components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/service/MediaService.java [hunk 4]
```diff
@@ -239,4 +244,55 @@ public void validateMediaType(String mediaTypePathParam, MediaType mediaType) {
...
+     * @param identifier File identifier.
+     * @return requested media file.
+     */
+    public Response downloadMediaFile(String type, String id, String identifier) {
+
+        // Retrieving a sub-representation of a media is not supported during the first phase of the implementation.
+        if (StringUtils.isNotBlank(identifier)) {
+            return Response.status(Response.Status.NOT_IMPLEMENTED).build();
+        }
+
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/service/MediaService.java:262]: shall we add a validation for id and return 404 if not valid?

File: components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/service/MediaService.java [hunk 5]
```diff
@@ -239,4 +244,55 @@ public void validateMediaType(String mediaTypePathParam, MediaType mediaType) {
...
+        if (resource instanceof FileContent) {
+            return Response.ok(((FileContent) resource).getFile()).header(HTTPConstants.HEADER_CONTENT_TYPE,
+                    ((FileContent) resource).getResponseContentType()).cacheControl(cacheControl).build();
+        } else if (resource instanceof StreamContent) {
+            return Response.ok().entity(((StreamContent) resource).getInputStream()).header(
+                    HTTPConstants.HEADER_CONTENT_TYPE, ((StreamContent) resource).getResponseContentType())
+                    .cacheControl(cacheControl).build();
+        }
+        return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/service/MediaService.java:276]: shall we add a a descriptive error message for the error scenario?

File: components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/service/MediaService.java [hunk 6]
```diff
@@ -239,4 +244,55 @@ public void validateMediaType(String mediaTypePathParam, MediaType mediaType) {
...
+    /**
+     * Download requested media file.
+     *
+     * @param type       The high level content-type of the resource (if media content-type is image/png then
+     *                   type would be image).
+     * @param id         Unique identifier for the requested media.
+     * @param identifier File identifier.
+     * @return requested media file.
+     */
+    public Response downloadMediaFile(String type, String id, String identifier) {
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/service/MediaService.java:256]: shall we add validation for type also? what will happen if type is null?

Comment (dewniMW) [reply components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/service/MediaService.java:256]: added a validation: https://github.com/wso2/identity-media/pull/19/files#diff-b318058c400489cbe88ca60afe7d6110R308

File: components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/common/MediaEndpointDataHolder.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+package org.wso2.carbon.identity.media.endpoint.common;
+
+import org.wso2.carbon.user.core.service.RealmService;
+
+/**
+ * OSGi service holder class for media service.
+ */
+public class MediaEndpointDataHolder {
+
+    private static RealmService realmService;
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/common/MediaEndpointDataHolder.java:26]: use singleton pattern for data holder and make these instance variables

File: components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/impl/UserApiServiceImpl.java [hunk 8]
```diff
@@ -48,7 +48,7 @@ public Response privilegedUserListMediaInformation(String type, String id) {
     public Response privilegedUserUploadMedia(String type, List<InputStream> filesInputStream,
                                               List<Attachment> filesDetail, PrivilegedUserMetadata metadata) {
 
-        mediaService.validateMediaType(type, filesDetail.get(0).getContentType());
+        mediaService.validateAllowedMediaTypesForUploading(type, filesDetail.get(0).getContentType());
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.media.endpoint/src/main/java/org/wso2/carbon/identity/media/endpoint/impl/UserApiServiceImpl.java:51]: shall we change the method name to, validateFileUploadMediaTypes or something similar?

**Review by emswbandara — APPROVED** [review]

**Pull request merged**